configurablecare/app/src/main/java/com/google/fhir/examples/configurablecare/care/TaskManager.kt

```diff
@@ -0,0 +1,190 @@
+  }
+
+  suspend fun fetchQuestionnaireFromTaskLogicalId(taskResourceId: String): Questionnaire? {
+    val task = fhirEngine.get(ResourceType.Task, taskResourceId) as Task
+    val questionnaires =
+      fhirEngine.search<Questionnaire> {
+        filter(
+          Questionnaire.IDENTIFIER,
+          { value = of(task.focus.reference.substring("Questionnaire/".length)) }
+        )
+      }
+    return questionnaires.firstOrNull()
+  }
+
+  suspend fun fetchServiceRequestFromTaskLogicalId(taskResourceId: String): ServiceRequest {
+    val task = fhirEngine.get(ResourceType.Task, taskResourceId) as Task
+    return fhirEngine.get(
+      ResourceType.ServiceRequest,
+      task.focus.reference.substring("ServiceRequest/".length)
+    ) as ServiceRequest
+  }
+
+  suspend fun getTasksForPatient(
+    patientId: String,
+    extraFilter: (Search.() -> Unit)?,
+    sort: (Search.() -> Unit)?
+  ): List<Task> {
+    return fhirEngine.search {
+      filter(Task.SUBJECT, { value = "Patient/$patientId" })
+      if (extraFilter != null) {
+        extraFilter()
+      }
+      operation = Operation.AND
+      if (sort != null) {
+        sort()
+      }
+    }
+  }
+
+  private suspend fun assignRequester(resource: Task, requesterId: String) {
+    val requesterResource =
+      fhirEngine.get(
+        ResourceType.fromCode(requesterId.substringBefore("/")),
+        IdType(requesterId).idPart
+      )
+    val requesterReference = Reference(requesterResource)
+    if (requesterResource is PractitionerRole) {
+      requesterReference.apply {
+        display = requesterResource.specialty.first()?.coding?.first()?.display
+      }
+    }
+    resource.apply { requester = requesterReference }
+  }
+
+  override suspend fun assignOwner(resource: Task, ownerId: String) {
+    val ownerResource =
+      fhirEngine.get(ResourceType.fromCode(ownerId.substringBefore("/")), IdType(ownerId).idPart)
+    val ownerReference = Reference(ownerResource)
+    if (ownerResource is PractitionerRole) {
+      ownerReference.apply { display = ownerResource.specialty.first()?.coding?.first()?.display }
+    }
+    resource.apply { owner = ownerReference }
+  }
+
+  suspend fun getTasksCount(patientId: String, extraFilter: (Search.() -> Unit)?): Int {
+    return extraFilter?.let { getTasksForPatient(patientId, it, null).count() } ?: 0
+  }
+
+  fun createTrackingTaskForServiceRequest(
+    serviceRequest: ServiceRequest,
+    configMap: MutableMap<String, String>,
+    subjectReference: Reference,
+    description: String?
+  ): Task {
+    val task = Task()
+    task.owner = serviceRequest.requester
+    task.description = "[Tracking Task] $description"
+    task.`for` = subjectReference
+    task.focus = Reference("ServiceRequest/${serviceRequest.id}")
+    task.status = TaskStatus.READY
+    task.intent = Task.TaskIntent.ORDER
+    task.lastModified = Date.from(Instant.now())
+
+    val serviceRequestFieldMap = ConfigurationManager.getServiceRequestConfigMap()
```

Comment:
Let's not use ConfigurationManager here. If we want TaskManager to have access to ServiceRequestConfig then we can pass it as a parameter.

I understand that maxDuration and unit of this tracking task is being set to that of ServiceRequests'.
But shouldn't we be using that from taskConfigMap ?

Reply:
This implementation has changed so that we use requestResourceConfiguration. 
To your point about maxDuration - we need to use the ServiceRequest's config because the Task is supposed to track its progress as opposed to being a standalone screening task for example